core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala

```diff
@@ -226,6 +259,135 @@ class HistoryServer(
 }
 
 /**
+ * An abstract implementation of the metrics [[Source]] trait with some common operations.
+ */
+private[history] abstract class HistoryMetricSource(val prefix: String) extends Source {
```

Comment:
done; also made the`Timestamp` gauge take a Spark `Clock` for easier use/testing, plus some other standalone tests. Separate source file -> isolated tests.

These gauges are actually generic to any metrics, and the `LambdaLongGauge` gauge quite a nice one to use in a lot of places; if accompanied with `LambdaIntGauge` they could be used as a more functional equivalent to the various gauge subclasses throughout the code. I could put them somewhere more generic if you think its useful (and even, in a different patch, move the existing gauges over to them)